### ISSUE TITLE
Traktor Kontrol S2 MK3 Hotcue LED issue

### ISSUE THREAD

**Issue by Joskagyerek3000** [mixxxdj/mixxx#12130]
### Bug Description

With the built-in mapping for the Traktor Kontrol S2 MK3, the hotcue pads only light up or change colors when I press the "HOTCUES" switch button. Otherwise, they don't give any feedback. 
In other words, it seems like that the 'HOTCUES' button makes a refresh to the hotcue LEDs, otherwise they don't show any changes in light or color.

### Version

2.4.0

### OS

Windows 11

**Comment by Swiftb0y** [comment]
Hey there, thank you for filling your issue. In order to get to the bottom of this. We would like to inspect your log file. Please use [our guide to find it](https://github.com/mixxxdj/mixxx/wiki/Finding%20the%20Mixxx.log%20file) and then upload it here and we'll see if we can spot anything in the log. Thank you.

**Comment by Joskagyerek3000** [comment]
Hello, here is the log file, I hope it helps. If I can do anything, or You
need any more information, I am happy to help further!

Thanks!

Swiftb0y ***@***.***> ezt írta (időpont: 2023. okt. 17., K,
20:22):

**Comment by ronso0** [comment]
Attachments don't show up on Github if you reply via email.

**Comment by Joskagyerek3000** [comment]
[mixxx.log](https://github.com/mixxxdj/mixxx/files/12938418/mixxx.log)

sorry, here it is :)

**Comment by subtruth** [comment]
Same problem, same version, different OS (Lubuntu 24.04)

[mixxx.log](https://github.com/user-attachments/files/16045474/mixxx.log)

**Comment by Swiftb0y** [comment]
@acolombier or @ywwg any chance you can look into this?

**Comment by lc-guy** [comment]
Getting this on 2.5.0-beta (Windows 10) as well.

It seems to me that the s2 mk3 script only calls padModeHandler when a pad button is clicked, and that is the only function where the colors and lights are set for cues. Maybe it should be called when a hotcue is updated?

**Comment by ywwg** [comment]
I do not own this controller, unfortunately.  I don't know who else in mixxx has an s2

**Comment by Swiftb0y** [comment]
True. Maybe @mi01 (original author) or @olivier-mauras can help out?

**Comment by Swiftb0y** [comment]
unfortunately the issue seems to interact with the `common-hid-packet-parser.js` library which I'm unfamiliar with. The offending code is here:
https://github.com/mixxxdj/mixxx/blob/b64251f1301c36c364c69da9348cd068d0b995dc/res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js#L888-L894

The log file gives a pretty good indication of the cause: 
```
Critical [Controller] HIDController.setOutput - Unknown field: [Channel2].hotcue_1_enabled
Warning [Controller] HIDController.linkOutput - Output already linked: [Channel1]
Warning [Controller] HIDController.linkOutput - Output already linked: [Channel2]
```

Maybe @JoergAtGithub can have a look?

**Comment by JoergAtGithub** [comment]
This seems to be a regression of #4637. Colored hotcues are a new mapping feature of 2.4.0.

The message
```
Warning [Controller] HIDController.linkOutput - Output already linked: [Channel1]
```

are easy to explain, because `"[Channel1]", "pad_" + i` is linked twice in the loop. But I do not see the reason for the `Unknown field` error logged before yet.

**Comment by Swiftb0y** [comment]
Thanks for the explanation. Well @goddisignz may be able to fix the code then.

**Comment by goddisignz** [comment]
I will try to have a look at this next week.

**Comment by Swiftb0y** [comment]
Thanks :+1: :+1: :+1:

**Comment by lc-guy** [comment]
@goddisignz were you able to look into it in the end?

**Comment by JeremyDeJong** [comment]
Bumping this one, I am happy to take a look sometime.

**Comment by Swiftb0y** [comment]
@JeremyDeJong that would be great. Can you reproduce the issue? If so can you try fixing it and open a PR?

**Comment by Terka96** [comment]
Hey, I've just found where is the problem.

In file /res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js callback functions hotcueOutputHandler and hotcueColorHandler calls colorOutputHandler and outputHandler with "key" parameter with value "hotcue_X_enabled"/"hotcue_X_color" instead of "pad_X".

Fix that worked for me was to create in both hotcue handler functions controllerKey by using string replace method and use it while calling colorOutputHandler and outputHandler:

in hotcueOutputHandler:
`  const controllerKey = key.replace("_enabled","").replace("hotcue_","pad_"); `
in hotcueColorHandler:
`  const controllerKey = key.replace("_color","").replace("hotcue_","pad_"); `

and replace "key" variable in this method calls with "controllerKey" like that:
` TraktorS2MK3.colorOutputHandler(color, group, controllerKey); `
` TraktorS2MK3.outputHandler(0, group, controllerKey); `


Also while switching pad mode by default pads in hotcue mode lights up with full brightness and it can be fixed by adding else case to if(active) statement in method padModeHandler:

`  TraktorS2MK3.outputHandler(0, field.group, "pad_" + i); `

I've tested the solution on my controller and it seems to work. I don't know what is the cleanest way to apply this fixes and I don't have now more time so I won't make PR.
